### DISCUSSION TITLE
Loading external stylesheets

### DISCUSSION THREAD

**Discussion by petewarman** [vercel/next.js#50796] [App Router]
My organisation has a set of shared styles that are used across all our various front-end applications, and we load them in from a CDN so our users get the benefit of having that asset cached as they move around the site. Importing these styles at build time means each individual application will create their own hashed version of the assets and therefore each subsequent page load will be slower as a result.

According to the app-router docs at the moment this isn't possible. However, it seems like doing something like this in the root layout does work but isn't documented

```
return (
  <html>
    <head>
      <link rel="stylesheet" href="/path/to/styles.css" />
    </head>
    <body>{children}</body>
  </html>
)
```

Are there plans to document this or provide an alternative approach?

**Comment by gerardnico** [comment]
If the stylesheets are not dynamic, I would say that this is the way to go.
Otherwise:
- there is also [useInsertionEffect](https://react.dev/reference/react/useInsertionEffect) to load dynamically a style script.
- may be that [useEffect](https://react.dev/reference/react/useEffect) can also do the job

**Reply by petewarman** [reply]
I don't think useEffect or useInsertionEffect would work for us. That would delay the loading of the styles until after the JavaScript has downloaded and executed, so we'd probably end up with a layout shift issue. These are core page styles that need to be part of the server rendered response.

**Reply by gerardnico** [reply]
Just an idea:

Did you try to use this minimal layout?
You may then add the head and body from within the page.js file?

```jsx
export default function RootLayout({ children }: { children: React.ReactNode; }) {
    return (
        <html lang="en">
        {children}
        </html>
    );
}
```

**Comment by devjiwonchoi** [comment]
You need to manipulate the ReactDom to safely insert into the `head` tag.

> you can use new [ReactDOM methods](https://github.com/facebook/react/pull/26237) to safely insert them into the <head> of the document.

```js
// app/preload-resources.ts
'use client'

import ReactDOM from 'react-dom'

export function PreloadResources() {
  ReactDOM.preload('CDN_CSS_HREF', { as: 'style' })

  return null
}
```

Note that the option `as` needs to be one of these: `'font' | 'script' | 'style'`

## Reference
- [Resource hints](https://nextjs.org/docs/app/api-reference/functions/generate-metadata#resource-hints)

**Comment by blopa** [comment]
Hey everyone,

I’m running into a bit of a snag and was hoping someone here might have some ideas. I’ve got this page where, based on certain data, it needs to pull in different external CSS. The tricky part is that the URL for the CSS is always changing, so I can’t set it up before rendering  (SSR or SSG), it needs to be dynamic.

### **Problem Summary:**
I’m getting warnings every time I try to inject this dynamic external CSS. I’ve tried the usual suspects like adding a `<link>` tag into the head of the document and using `dangerouslySetInnerHTML`, but no luck so far.

I also tried the [Next.js Document](https://nextjs.org/docs/pages/building-your-application/routing/custom-document), but since that is only SSR, it didn't work - I tried hooking it into a context to change the CSS for each dynamic URL but nothing.

Like, I know I can manually change the header using some vanilla JS trickery and stuff, and also using `react-helmet` seems to work, but I was looking for a native Next.js solution for it.

### **Questions:**
1. Anyone know of a way to dynamically add external CSS to a Next.js page without the warnings, especially when dealing with a URL that won’t sit still?
2. Any clever workarounds or other methods to get around the warnings and still pull in the dynamic external CSS?
3. Anyway to do it without external libraries?

The Next.js version I'm using is 13.4.19.

Any thoughts or suggestions would be super appreciated!

Thanks a bunch in advance for any help or insights.

**Reply by mutou981** [reply]
@blopa You can write the link element directly in the body.

```tsx
<html lang={locale} suppressHydrationWarning>
      <body
        className={cn(
          "min-h-screen bg-background font-sans antialiased",
          outfit.variable,
        )}
      >
        {locale === "zh-CN" && (
          <link
            rel="stylesheet"
            crossOrigin="anonymous"
            href="https://cdn.jsdelivr.net/npm/misans@4.0.0/lib/Normal/MiSans-Regular.min.css"
          />
        )}
          <Providers>{children}</Providers>
      </body>
</html>
```
